test_runner/src/main/kotlin/ftl/client/google/run/ios/GcIosTestMatrix.kt

```diff
@@ -0,0 +1,93 @@
+import ftl.util.timeoutToSeconds
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.coroutineScope
+
+suspend fun executeIosTests(
+    config: TestMatrixIos.Config,
+    testMatrixTypes: List<TestMatrixIos.Type>,
+): List<TestMatrix> = testMatrixTypes
+    .foldIndexed(emptyList<Deferred<TestMatrix>>()) { _, testMatrices, testMatrixType ->
+        testMatrices + executeIosTestMatrixAsync(testMatrixType, config)
+    }.awaitAll()
+
+private suspend fun executeIosTestMatrixAsync(
+    type: TestMatrixIos.Type,
+    config: TestMatrixIos.Config
+): Deferred<TestMatrix> = coroutineScope {
+    async(Dispatchers.IO) {
+        createIosTestMatrix(type, config).executeWithRetry()
+    }
+}
+
+private fun createIosTestMatrix(
+    testMatrixType: TestMatrixIos.Type,
+    config: TestMatrixIos.Config,
+): Testing.Projects.TestMatrices.Create {
+
+    val testMatrix = TestMatrix()
+        .setClientInfo(config.clientInfo)
+        .setTestSpecification(getIosTestSpecification(testMatrixType, config))
+        .setEnvironmentMatrix(config.environmentMatrix)
+        .setResultStorage(resultsStorage(config, testMatrixType))
+        .setFlakyTestAttempts(config.flakyTestAttempts)
+        .setFailFast(config.failFast)
+
+    return runCatching {
+        GcTesting.get.projects().testMatrices().create(config.project, testMatrix)
+    }.getOrElse { throw FlankGeneralError(it) }
+}
+
+private fun resultsStorage(config: TestMatrixIos.Config, type: TestMatrixIos.Type): ResultStorage {
+    return ResultStorage().setGoogleCloudStorage(
+        GoogleCloudStorage().setGcsPath(type.matrixGcsPath)
+    ).setToolResultsHistory(
+        ToolResultsHistory().setHistoryId(config.resultsHistoryName).setProjectId(config.project)
+    )
+}
+
+private val TestMatrixIos.Config.environmentMatrix
+    get() = EnvironmentMatrix().setIosDeviceList(GcIosDevice.build(devices))
+
+private fun getIosTestSpecification(
+    testMatrixType: TestMatrixIos.Type,
+    config: TestMatrixIos.Config
+): TestSpecification = TestSpecification()
+    .setDisableVideoRecording(!config.recordVideo)
+    .setTestTimeout("${timeoutToSeconds(config.testTimeout)}s")
+    .setIosTestSetup(getIosTestSetup(config))
+    .setupIosTest(testMatrixType)
+
+private fun getIosTestSetup(
+    config: TestMatrixIos.Config
+): IosTestSetup = IosTestSetup()
+    .setNetworkProfile(config.networkProfile)
+    .setPushFiles(config.otherFiles.mapToIosDeviceFiles())
+    .setAdditionalIpas(config.additionalIpasGcsPaths.mapGcsPathsToFileReference())
+    .setPullDirectories(config.directoriesToPull.toIosDeviceFiles())
+
+private fun List<String>.toIosDeviceFiles() = map { path -> toIosDeviceFile(path) }
+
+private val TestMatrixIos.Config.clientInfo
+    get() = ClientInfo().setName("Flank").setClientInfoDetails(clientDetails?.toClientInfoDetailList())
```

Comment:
some of these functions are common for Android and iOS could we share them?

Reply:
@piotradamczyk5 Sure can look into it :)

Reply:
There is now differences in androids clientinfo. I dont see a reason to try merge it.